src/trans.rs

```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
                     let expr = unsafe { BinaryenReturn(self.module, expr) };
                     binaryen_stmts.push(expr);
                 }
+                TerminatorKind::Switch { ref discr, .. } => {
+                    let adt = self.trans_lval(discr);
+                    let adt_ty = self.mir.lvalue_ty(*self.tcx, discr).to_ty(*self.tcx);
+
+                    if adt.offset.is_some() {
+                        panic!("unimplemented Switch with offset");
+                    }
+
+                    let adt_layout = self.type_layout(adt_ty);
+                    let discr_val = match *adt_layout {
+                        Layout::General { discr, .. } => {
+                            let discr_size = discr.size().bytes() as u32;
+                            debug!("emitting GetLocal({}) + Load for ADT Switch condition", adt.index.0);
+                            unsafe {
+                                let ptr = BinaryenGetLocal(self.module, adt.index, BinaryenInt32());
+                                BinaryenLoad(self.module, discr_size, 0, 0, 0, BinaryenInt32(), ptr)
+                            }
+                        }
+                        Layout::CEnum { .. } => {
```

Comment:
Just to make sure I understand this right, CEnum layout means the enum is just a single value, and therefore passed by value instead of as a pointer to the tagged union?